### ISSUE TITLE
nixos-rebuild build-vm-with-bootloader fails

### ISSUE THREAD

**Issue by matthiasbeyer** [NixOS/nixpkgs#8666]
I can't see why, but maybe you can.

```
[...]
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/systemd-udevd
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/e2fsck
'/nix/store/ybkwnkwchv043b1fgj0csrifcx044asi-e2fsprogs-1.42.13/lib/libext2fs.so.2' -> '/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib/libext2fs.so.2'
'/nix/store/ybkwnkwchv043b1fgj0csrifcx044asi-e2fsprogs-1.42.13/lib/libext2fs.so.2.4' -> '/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib/libext2fs.so.2.4'
'/nix/store/ybkwnkwchv043b1fgj0csrifcx044asi-e2fsprogs-1.42.13/lib/libcom_err.so.2' -> '/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib/libcom_err.so.2'
'/nix/store/ybkwnkwchv043b1fgj0csrifcx044asi-e2fsprogs-1.42.13/lib/libcom_err.so.2.1' -> '/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib/libcom_err.so.2.1'
'/nix/store/ybkwnkwchv043b1fgj0csrifcx044asi-e2fsprogs-1.42.13/lib/libe2p.so.2' -> '/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib/libe2p.so.2'
'/nix/store/ybkwnkwchv043b1fgj0csrifcx044asi-e2fsprogs-1.42.13/lib/libe2p.so.2.3' -> '/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib/libe2p.so.2.3'
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/tune2fs
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/blkid
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/ata_id
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/dmsetup
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/mke2fs
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/kmod
Copying libs for bin /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/busybox
'/nix/store/i0l0jjkk82wsqz9z5yhg35iy78bjq684-glibc-2.21/lib/libm.so.6' -> '/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib/libm.so.6'
'/nix/store/i0l0jjkk82wsqz9z5yhg35iy78bjq684-glibc-2.21/lib/libm-2.21.so' -> '/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib/libm-2.21.so'
stripping (with flags -s) in /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/lib  /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin 
building path(s) ‘/nix/store/w8dnf804a3gd9r9qv0jp4z5mx9zf8jwz-unit-script’
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/ata_id...
warning: working around a Linux kernel bug by creating a hole of 4096 bytes in ‘/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/ata_id’
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/blkid...
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/busybox...
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/cdrom_id...
warning: working around a Linux kernel bug by creating a hole of 4096 bytes in ‘/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/cdrom_id’
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/dmsetup...
warning: working around a Linux kernel bug by creating a hole of 2101248 bytes in ‘/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/dmsetup’
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/e2fsck...
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/kmod...
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/lvm...
warning: working around a Linux kernel bug by creating a hole of 2215936 bytes in ‘/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/lvm’
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/mdadm...
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/mke2fs...
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/scsi_id...
warning: working around a Linux kernel bug by creating a hole of 4096 bytes in ‘/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/scsi_id’
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/systemd-udevd...
warning: working around a Linux kernel bug by creating a hole of 4096 bytes in ‘/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/systemd-udevd’
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/tune2fs...
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/udevadm...
warning: working around a Linux kernel bug by creating a hole of 4096 bytes in ‘/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/udevadm’
patching /nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/v4l_id...
warning: working around a Linux kernel bug by creating a hole of 4096 bytes in ‘/nix/store/iqcmbshpccl18plbsigswszx8x58ny6w-extra-utils/bin/v4l_id’
testing patched programs...
hello world
217
mdadm - v3.3 - 3rd September 2013
11253 blocks
building path(s) ‘/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules’
building path(s) ‘/nix/store/imh98s5s1xjgm2cm55jf9pwmhc74lfv8-vm-run’
'/nix/store/zlq94gj8mdb1vxpvwh7vzjm6c24y75rz-systemd-217/lib/udev/rules.d/60-cdrom_id.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/60-cdr
'/nix/store/zlq94gj8mdb1vxpvwh7vzjm6c24y75rz-systemd-217/lib/udev/rules.d/60-persistent-storage.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-ru
'/nix/store/zlq94gj8mdb1vxpvwh7vzjm6c24y75rz-systemd-217/lib/udev/rules.d/80-drivers.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/80-driv
'/nix/store/5kaiqpwwnsqycqf5qlbx2q042lqkb5v6-lvm2-2.02.118/lib/udev/rules.d/10-dm.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/10-dm.rule
'/nix/store/5kaiqpwwnsqycqf5qlbx2q042lqkb5v6-lvm2-2.02.118/lib/udev/rules.d/11-dm-lvm.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/11-dm-
'/nix/store/5kaiqpwwnsqycqf5qlbx2q042lqkb5v6-lvm2-2.02.118/lib/udev/rules.d/13-dm-disk.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/13-dm
'/nix/store/5kaiqpwwnsqycqf5qlbx2q042lqkb5v6-lvm2-2.02.118/lib/udev/rules.d/95-dm-notify.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/95-
'/nix/store/6cdycnwk3brq805z49k2wzmnbd2qx99z-mdadm-3.3/lib/udev/rules.d/63-md-raid-arrays.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/63
'/nix/store/6cdycnwk3brq805z49k2wzmnbd2qx99z-mdadm-3.3/lib/udev/rules.d/64-md-raid-assembly.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/
'/nix/store/7y74n8rdm5y4k0k2ivj7pcd10bijxp80-bcache-tools-1.0.7/lib/udev/rules.d/69-bcache.rules' -> '/nix/store/nafhwjpbj47m0zskcwkgd9mja1c5p02x-udev-rules/6
building path(s) ‘/nix/store/hy2c1m6bjvdc5s49y6fa240zj84j68bx-unit-cron.service’
building path(s) ‘/nix/store/n5ijhy8152cwkv96madv4bnjpxicmfzk-stage-1-init.sh’
building path(s) ‘/nix/store/5hq0p3y2rfgl97gz74lmm2wgz7h45wkj-system-units’
building path(s) ‘/nix/store/2k555j8sfdzl7ly115w9lss9229j0xya-initrd’
building path(s) ‘/nix/store/f0rbzj7xv0rd1n1z9hcrlf7f1g82iwph-etc’
37559 blocks
building path(s) ‘/nix/store/ymgbh1qv17zlwzfbw19ln83m19mxjmjw-nixos-15.06pre64030.e1af50c’
building path(s) ‘/nix/store/dkm4fz9j8xfw2lvapirxwpggmqdyjxyv-nixos-boot-disk’
Formatting '/nix/store/dkm4fz9j8xfw2lvapirxwpggmqdyjxyv-nixos-boot-disk/disk.img', fmt=qcow2 size=41943040 encryption=off cluster_size=65536 lazy_refcounts=of
QEMU 2.2.1 monitor - type 'help' for more information
(qemu) char device redirected to /dev/pts/0 (label serial0)
builder for ‘/nix/store/51v99i0a8ci7cwpymrkbpya0xfw3qcbj-nixos-boot-disk.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/h1a9n5hdz345iz90l5rwc25mi660v3sz-run-nixos-vm.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/dwsjgsslsyzbijy56wlm3cv5lykxwfai-nixos-vm.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/dwsjgsslsyzbijy56wlm3cv5lykxwfai-nixos-vm.drv’ failed
```

**Comment by matthiasbeyer** [comment]
Maybe cc @lethalman

**Comment by lethalman** [comment]
Built with -j1 and repaste.

**Comment by matthiasbeyer** [comment]
```
$ sudo nixos-rebuild build-vm-with-bootloader -j 1
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/51v99i0a8ci7cwpymrkbpya0xfw3qcbj-nixos-boot-disk.drv
  /nix/store/h1a9n5hdz345iz90l5rwc25mi660v3sz-run-nixos-vm.drv
  /nix/store/dwsjgsslsyzbijy56wlm3cv5lykxwfai-nixos-vm.drv
building path(s) ‘/nix/store/dkm4fz9j8xfw2lvapirxwpggmqdyjxyv-nixos-boot-disk’
Formatting '/nix/store/dkm4fz9j8xfw2lvapirxwpggmqdyjxyv-nixos-boot-disk/disk.img', fmt=qcow2 size=41943040 encryption=off cluster_size=65536 lazy_refcounts=off 
QEMU 2.2.1 monitor - type 'help' for more information
(qemu) char device redirected to /dev/pts/0 (label serial0)
builder for ‘/nix/store/51v99i0a8ci7cwpymrkbpya0xfw3qcbj-nixos-boot-disk.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/h1a9n5hdz345iz90l5rwc25mi660v3sz-run-nixos-vm.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/dwsjgsslsyzbijy56wlm3cv5lykxwfai-nixos-vm.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/dwsjgsslsyzbijy56wlm3cv5lykxwfai-nixos-vm.drv’ failed
```

**Comment by telent** [comment]
I get exactly the same, modulo that it's a newer version of qemu.  Also, it hangs for a couple of seconds after printing `(qemu) char device redirected to /dev/pts/2 (label serial0)`
```
[dan@noetbook:~]$ nixos-rebuild build-vm-with-bootloader -j 1
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/2hsygpqg8vklyldjkn74p1hn24wxhbk2-nixos-boot-disk.drv
  /nix/store/lbq3b9f6m1c9kz65aw7hqs8xhxv51x1q-run-nixos-vm.drv
  /nix/store/82kjh0mizm57h0wj86m9rpsd7p9c3sp1-nixos-vm.drv
building path(s) ‘/nix/store/h23payspm1ic7kanjp6sq3ygvi5niya1-nixos-boot-disk’
Formatting '/nix/store/h23payspm1ic7kanjp6sq3ygvi5niya1-nixos-boot-disk/disk.img', fmt=qcow2 size=41943040 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
QEMU 2.4.0 monitor - type 'help' for more information
(qemu) char device redirected to /dev/pts/2 (label serial0)
builder for ‘/nix/store/2hsygpqg8vklyldjkn74p1hn24wxhbk2-nixos-boot-disk.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/lbq3b9f6m1c9kz65aw7hqs8xhxv51x1q-run-nixos-vm.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/82kjh0mizm57h0wj86m9rpsd7p9c3sp1-nixos-vm.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/82kjh0mizm57h0wj86m9rpsd7p9c3sp1-nixos-vm.drv’ failed
```

This is on NixOS 15.09.897.aa2d714, as installed via usb from the "minimal" image.  Any ideas?

**Comment by zimbatm** [comment]
Just tested it and it's still an issue.

**Comment by zimbatm** [comment]
The following change fixes the build for me but I don't really understand what's going on here (the second change is because sfdisk can only toggle the bootable flag on MBR disk layouts). Why is the script trying to build a EFI partition even if `cfg.useEFIBoot` is set to false. Also how do you use the end-result ?

```
diff --git a/nixos/modules/virtualisation/qemu-vm.nix b/nixos/modules/virtualisation/qemu-vm.nix
index 7dfbc38..58d34c6 100644
--- a/nixos/modules/virtualisation/qemu-vm.nix
+++ b/nixos/modules/virtualisation/qemu-vm.nix
@@ -128,8 +128,8 @@ let
             '';
           buildInputs = [ pkgs.utillinux ];
           QEMU_OPTS = if cfg.useEFIBoot
-                      then "-pflash $out/bios.bin -nographic -serial pty"
-                      else "-nographic -serial pty";
+                      then "-pflash $out/bios.bin -nographic"
+                      else "-nographic";
         }
         ''
           # Create a /boot EFI partition with 40M
@@ -140,7 +140,6 @@ let
           ${pkgs.gptfdisk}/sbin/sgdisk -A 2:set:2 /dev/vda
           ${pkgs.gptfdisk}/sbin/sgdisk -h 2 /dev/vda
           ${pkgs.gptfdisk}/sbin/sgdisk -C /dev/vda
-          ${pkgs.utillinux}/bin/sfdisk /dev/vda -A 2
           . /sys/class/block/vda2/uevent
           mknod /dev/vda2 b $MAJOR $MINOR
           . /sys/class/block/vda/uevent
```

**Comment by matthiasbeyer** [comment]
Closing because too old to get addressed IMO.

**Comment by bjornfor** [comment]
I think this was fixed in nixos-20.03. At least I can do this at commit 51d115ac89d676345b05a0694b23bd2691bf708a, which I couldn't on nixos-19.09:

```
$ nixos-rebuild build-vm-with-bootloader; echo $?
building Nix...
        building the system configuration...
0
```
